engines/azure/app/models/azure/interface.rb

```diff
@@ -0,0 +1,358 @@
+module Azure
+  require 'base64'
+
+  class Interface
+    class LoginError < StandardError; end
+
+    RESOURCE = 'https://management.azure.com/'
+    HOST = 'https://management.azure.com'
+
+    attr_accessor(:credential, :subscription, :region, :token)
+
+    def self.load
+      new(
+        credential: Azure::Credential.last,
+        subscription: Azure::Subscription.load(),
+        region: Azure::Region.load()
+      )
+    end
+
+    def initialize(credential:, subscription:, region:)
+      @credential = credential
+      @subscription = subscription
+      @region = region
+    end
+
+    def login(credentials: @credential.creation_attributes)
+      if @subscription.value.blank?
+        @subscription.value = Azure::Metadata.get_subscription()
+        @subscription.save!
+      end
+      @token = get_token(**credentials)
+    rescue RestClient::ExceptionWithResponse => e
+      raise(LoginError.new(
+        JSON.parse(e.response.body)['error_description']
+      ))
+    rescue StandardError => e
+      raise(LoginError.new(e.message))
+    end
+
+    def get_token(app_id:, password:, tenant:)
+      response = RestClient.post(
+        tenant_oauth_token_url(tenant: tenant),
+        {
+          'grant_type' => 'client_credentials',
+          'client_id' => app_id,
+          'client_secret' => password,
+          'resource' => RESOURCE
+        },
+        {
+          accept: :json
+        }
+      )
+      # TODO capture expiration time
+      JSON.parse(response)['access_token']
+    end
+
+    def tenant_oauth_token_url(tenant:)
+      "https://login.microsoftonline.com/#{ tenant }/oauth2/token"
+    end
+
+    def api_exception_trap(capture_exception)
+      yield
+    rescue RestClient::ExceptionWithResponse => e
+      raise(e) unless capture_exception
+      e.response
+    end
+
+    # REST wrappers
+
+    def api_get(path:, params: {}, capture_exception: false)
+      self.login() if @token.blank?
+      response = api_exception_trap(capture_exception) do
+        RestClient.get(
+          HOST + "/subscriptions/#{@subscription.value}" + path,
+          {
+            params: {
+              'api-version' => '2023-07-01',
```

Comment:
could we use a const for the repeating api version values ?

Reply:
How about 63ade51 ?